src/main/java/ch/njol/skript/expressions/ExprMaxStack.java

```diff
@@ -1,62 +1,118 @@
-import org.bukkit.Material;
-import org.bukkit.inventory.ItemStack;
+import ch.njol.util.Math2;
+import ch.njol.util.coll.CollectionUtils;
 
-/**
- * @author joeuguce99
- */
 @Name("Maximum Stack Size")
-@Description("The maximum stack size of the specified material, e.g. 64 for torches, 16 for buckets, and 1 for swords.")
-@Examples("send \"You can only pick up %max stack size of player's tool% of %type of (player's tool)%\" to player")
-@Since("2.1")
-public class ExprMaxStack extends SimplePropertyExpression<ItemType, Long> {
+@Description({
+	"The maximum stack size of an item (e.g. 64 for torches, 16 for buckets, 1 for swords, etc.) or inventory.",
+	"In 1.20.5+, the maximum stack size of items can be changed to any integer from 1 to 99, and stacked up to the maximum stack size of the inventory they're in."
+})
+@Examples({
+	"send \"You can hold %max stack size of player's tool% of %type of player's tool% in a slot.\" to player",
+	"set the maximum stack size of inventory of all players to 16",
+	"add 8 to the maximum stack size of player's tool",
+	"reset the maximum stack size of {_gui}"
+})
+@Since("2.1, INSERT VERSION (changeable, inventories)")
+@RequiredPlugins("Spigot 1.20.5+ (changeable)")
+public class ExprMaxStack extends SimplePropertyExpression<Object, Integer> {
 
 	static {
-		register(ExprMaxStack.class, Long.class, "max[imum] stack[[ ]size]", "itemtype");
+		register(ExprMaxStack.class, Integer.class, "max[imum] stack[[ ]size]", "itemtypes/inventories");
 	}
-	
-	@SuppressWarnings("null")
+
+	private static final boolean CHANGEABLE_ITEM_STACK_SIZE = Skript.methodExists(ItemMeta.class, "setMaxStackSize", Integer.class);
+
 	@Override
-	public Long convert(ItemType itemType) {
-		Object random = itemType.getRandomStackOrMaterial();
-		if (random instanceof Material)
-			return (long) ((Material) random).getMaxStackSize();
-		return (long) ((ItemStack) random).getMaxStackSize();
+	public @Nullable Integer convert(Object from) {
+		if (from instanceof ItemType itemType)
+			return getMaxStackSize(itemType);
+		if (from instanceof Inventory inventory)
+			return inventory.getMaxStackSize();
+		return null;
 	}
 
 	@Override
-	public Class<? extends Long> getReturnType() {
-		return Long.class;
+	public @Nullable Class<?>[] acceptChange(ChangeMode mode) {
+        return switch (mode) {
+            case ADD, REMOVE, RESET, SET -> {
+                if (!CHANGEABLE_ITEM_STACK_SIZE && ItemType.class.isAssignableFrom(getExpr().getReturnType())) {
+                    Skript.error("Changing the maximum stack size of items requires Minecraft 1.20.5 or newer!");
+                    yield null;
+                }
+                yield CollectionUtils.array(Integer.class);
+            }
+            default -> null;
+        };
+	}
+
+	@Override
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
+		Integer change = null;
+		if (mode != ChangeMode.RESET)
+			change = (int) delta[0];
```

Comment:
This part is kinda weird to me. I usually just do this instead, which would mean some of the following logic would have to change too (mainly just where you check if `change != null` I think)
```suggestion
		int change = delta == null ? 0 : ((Number) delta[0]).intValue();
```